lib/compiler-cranelift/src/translator/func_translator.rs

```diff
@@ -256,6 +262,29 @@ fn parse_function_body<FE: FuncEnvironment + ?Sized>(
     if state.reachable {
         debug_assert!(builder.is_pristine());
         if !builder.is_unreachable() {
+            // TODO: encapsulate this logic into `environ` to avoid needing to allocate a `Vec` for the borrow checker
+            for (local_index, local_type) in environ.get_local_types().to_vec().iter().enumerate() {
+                if *local_type == WasmerType::ExternRef {
+                    let val = builder.use_var(Variable::with_u32(local_index as _));
+                    environ.translate_externref_dec(builder.cursor(), val)?;
+                }
+            }
+
+            let num_elems_to_drop = state.stack.len() - builder.func.signature.returns.len();
+            // drop elements on the stack that we're not returning
+            for val in state
+                .stack
+                .iter()
+                .zip(state.metadata_stack.iter())
+                .take(num_elems_to_drop)
+                .filter(|(_, metadata)| metadata.ref_counted)
+                .map(|(val, _)| val)
+            {
+                environ.translate_externref_dec(builder.cursor(), *val)?;
+            }
+
+            // TODO: look into what `state.reachable` check above does as well as `!builder.is_unreachable`, do we need that too for ref counting?
```

Comment:
This is a very good question. What happens to externref upon unreachable? Are they dec-ref'd?